src/Moq/ExpressionExtensions.cs

```diff
@@ -320,7 +320,9 @@ internal static PropertyInfo GetReboundProperty(this MemberExpression expression
 			// we "upgrade" to the derived property.
 			if (property.DeclaringType != expression.Expression.Type)
 			{
-				var derivedProperty = expression.Expression.Type.GetProperty(property.Name, BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance);
+				var derivedProperty = expression.Expression.Type
+					.GetProperties(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance)
+					.SingleOrDefault(p => p.Name == property.Name && p.PropertyType == property.PropertyType);
```

Comment:
Could you please rewrite this query as follows?:

```csharp
var derivedProperty = expression.Expression.Type
        .GetMember(property.Name, MemberTypes.Property, BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance)
        .Cast<PropertyInfo>()
        .SingleOrDefault(p => p.PropertyType == property.PropertyType);
```

This should be more efficient as the filtering by name happens earlier (inside System.Reflection).